src/crystal/system/unix/group.cr

```diff
@@ -0,0 +1,137 @@
+require "c/grp"
+require "c/sys/limits"
+
+struct Crystal::System::Group
+  {% if flag?(:darwin) || flag?(:openbsd) || flag?(:freebsd) %}
+    GID_MAX = LibC::GID_MAX
+  {% elsif flag?(:linux) %}
+    GID_MAX = 0xffffffff_u32
+  {% else %}
+    {{ raise "Unsupported platform, only Darwin, OpenBSD, FreeBSD, and Linux (GNU, musl) are supported." }}
+    #GID_MAX = 0xffff_u32 # POSIX Default
+  {% end %}
+  NGROUPS_MAX = LibC::NGROUPS_MAX
+
+  private GETGR_R_SIZE_MAX = begin
+    size = LibC.sysconf(LibC::SC_GETGR_R_SIZE_MAX)
+
+    # Set default size if sysconf not set
+    size < 0 ? 1024 : size
+  end
+
+  def self.getgrnam?(groupname : String) : LibC::Group?
+    groupname.check_no_null_byte
+
+    grp = uninitialized LibC::Group
+    buf = Bytes.new(GETGR_R_SIZE_MAX)
+    result = pointerof(grp)
+
+    ret = LibC.getgrnam_r(groupname, pointerof(grp), buf, buf.size, pointerof(result))
+
+    unless result
+      return if ret == 0 # not found
+      raise Errno.new("getgrnam_r")
+    end
+
+    grp
+  end
+
+  def self.getgrgid?(gid : String) : LibC::Group?
+    gid = to_gid(gid)
+    return nil if !gid
+    getgrgid?(gid)
+  end
+
+  def self.getgrgid?(gid : Int) : LibC::Group?
+    return if !valid_gid?(gid)
+
+    grp = uninitialized LibC::Group
+    buf = Bytes.new(GETGR_R_SIZE_MAX)
+    result = pointerof(grp)
+
+    ret = LibC.getgrgid_r(gid, pointerof(grp), buf, buf.size, pointerof(result))
+
+    unless result
+      return if ret == 0 # not found
+      raise Errno.new("getgrgid_r")
+    end
+
+    grp
+  end
+
+  def self.name?(gid : Int | String) : String?
+    grp = getgrgid?(gid)
+    return nil if !grp
+    String.new(grp.gr_name)
+  end
+
+  def self.gid?(groupname : String) : UInt32?
+    grp = getgrnam?(groupname)
+    return nil if !grp
+    grp.gr_gid
+  end
+
+  def self.get?(groupname : String) : self?
+    return nil if groupname == "-1"
+    gid = to_gid(groupname)
+    return get?(gid) if gid
+
+    grp = getgrnam?(groupname)
+    return nil if !grp
+    new(grp)
+  end
+
+  def self.get?(gid : Int) : self?
+    return nil if gid == -1
+
+    grp = getgrgid?(gid)
+    return nil if !grp
+    new(grp)
+  end
+
+  # :nodoc:
+  private def initialize(group : LibC::Group)
+    @name = String.new(group.gr_name)
+    @gid = group.gr_gid
+    slice = Slice.new(group.gr_mem, limit: NGROUPS_MAX, read_only: true)
+    @user_names = Array.new(slice.size) { |idx| String.new(slice[idx]) }
+  end
+
+  # Returns the group name as a `String`.
+  getter name : String
+
+  # Returns the group ID in a `String` representation.
+  def gid : String
+    return @gid.to_s
+  end
+
+  # Returns the group ID as a `UInt32`.
+  @gid : UInt32
+
+  def raw_gid : UInt32?
```

Comment:
This isn't needed, people can use `group.@gid`.

Reply:
I didn't realize you could directly access struct elements directly.